### ISSUE TITLE
Make UserDetailsServiceAutoConfiguration conditional on missing bean of type AuthenticationManagerResolver

### ISSUE THREAD

**Issue by SamuelBucheliZ** [spring-projects/spring-boot#26766]
When setting up an OAuth2 resource server with multi-tenancy (e.g., the one from the example https://github.com/spring-projects/spring-security-samples/tree/main/servlet/spring-boot/java/oauth2/resource-server/multi-tenancy ), none of the conditions for disabling the `UserDetailsServiceAutoConfiguration` are met and therefore, an InMemoryUserDetailsManager is auto-configured:
```
2021-06-01 18:34:39.832  INFO 26857 --- [           main] .s.s.UserDetailsServiceAutoConfiguration : 

Using generated security password: f89a4c05-3b75-4c2c-89a0-60dceba7e4ec
```

From my point of view, in this use case, the generation of a default passowrd is not desirable (I think a similar topic was addressed in PR #17646).

As an enhancement, I would therefore propose to also deactivate the `UserDetailsServiceAutoConfiguration` for such cases.

One way I see to achieve this, would be to make `UserDetailsServiceAutoConfiguration`conditional on missing beans of type `org.springframework.security.authentication.AuthenticationManagerResolver` as well. As far as I can tell, this would be consistent with the fact that `UserDetailsServiceAutoConfiguration` is already conditional on missing `AuthenticationManager` beans and, in some sense, having an `AuthenticationManagerResolver` is equivalent to having an `AuthenticationManager`. However, I have not fully investigated the implications of this yet. There may be better ways to achieve this.

**Comment by SamuelBucheliZ** [comment]
@mbhave Exactly, in the example, there are no beans for `JwtDecoder` or `OpaqueTokenIntrospector`. In the multi-tenancy example I linked above, the problem could be fixed by just adding a `@Bean` notation e.g., for the `JwtDecoder`, but this would be a bit of a hack.

However, in general, the problem cannot always be solved so easily. For example, in the use case of supporting multiple issuer URIs, one would typically use [JwtIssuerAuthenticationManagerResolver](https://docs.spring.io/spring-security/site/docs/current/api/org/springframework/security/oauth2/server/resource/authentication/JwtIssuerAuthenticationManagerResolver.html). In this case, obtaining a `JwtDecoder` bean is not as straightforward (respectively, creating one is still possible with a little bit of effort, but even more of a hack, as this bean wouldn't be used in any place).

For reference, the corresponding code (adapted from the example), would look something like this
```
@EnableWebSecurity
public class OAuth2ResourceServerSecurityConfiguration extends WebSecurityConfigurerAdapter {

	@Value("${tenantOne.issuer-uri}")
	String tenantOneIssuerUri;

	@Value("${tenantTwo.issuer-uri}")
	String tenantTwoIssuerUri;

	@Override
	protected void configure(HttpSecurity http) throws Exception {
		// @formatter:off
		http
			.authorizeRequests((requests) -> requests
					.mvcMatchers("/**/message/**").hasAuthority("SCOPE_message:read")
					.anyRequest().authenticated()
			)
			.oauth2ResourceServer((resourceServer) -> resourceServer
					.authenticationManagerResolver(multitenantAuthenticationManager())
			);
		// @formatter:on
	}

	@Bean
	AuthenticationManagerResolver<HttpServletRequest> multitenantAuthenticationManager() {
		return new JwtIssuerAuthenticationManagerResolver(
                  tenantOneIssuerUri,
                  tenantTwoIssuerUri
            );
	}

}
```

**Comment by jzheaux** [comment]
I agree that this bean supersedes creating a `UserDetailsService`.

Since the purpose of an `AuthenticationManagerResolver` is to produce an `AuthenticationManager`, it makes sense to me to apply the same conditional as is done for `AuthenticationManager`.

`ReactiveUserDetailsServiceAutoConfiguration` should likely be updated in the same way.

**Comment by nandishkotadia** [comment]
I am interested in picking this up. Can I work on the same?
@scottfrederick @mbhave

**Comment by qavid** [comment]
Hi @mbhave,
`ReactiveUserDetailsServiceAutoConfiguration` should be updated in the same way as already mentioned by @jzheaux.
Could you please reopen issue?

**Comment by philwebb** [comment]
@qavid I've opened a new issue to track this https://github.com/spring-projects/spring-boot/issues/31317
